tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/WithGlobals.scala

```diff
@@ -0,0 +1,95 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.core.tools.linker.backend.emitter
+
+import GlobalRefUtils.unionPreserveEmpty
+
+/** A monad that associates a set of global variable names to a value.
+ *
+ *  Definition:
+ *  {{{
+ *  WithGlobals.apply(a) === WithGlobals(a, ∅)
+ *
+ *  WithGlobals(v, w).flatMap(f) === WithGlobals(f(v).value, w ∪ f(v).globalVarNames)
+ *  }}}
+ *
+ *  Left identity:
+ *  {{{
+ *               WithGlobals.apply(a).flatMap(f)     =?= f(a)
+ *               --------------------
+ *           WithGlobals(a, ∅).flatMap(f)            =?= f(a)
+ *           ----------------------------
+ *  WithGlobals(f(a).value, ∅ ∪ f(a).globalVarNames) =?= f(a)
+ *                          -----------------------
+ *  WithGlobals(f(a).value,   f(a).globalVarNames  ) =?= f(a)
+ *  }}}
+ *
+ *  Right identity:
+ *  {{{
+ *  WithGlobals(v, w).flatMap(WithGlobals.apply)                                     =?= WithGlobals(v, w)
+ *  --------------------------------------------
+ *  WithGlobals(WithGlobals.apply(v).value, w ∪ WithGlobals.apply(v).globalVarNames) =?= WithGlobals(v, w)
+ *              --------------------------      -----------------------------------
+ *  WithGlobals(          v               , w ∪                 ∅                    =?= WithGlobals(v, w)
+ *  }}}
+ *
+ *  Associativity:
+ *  {{{
+ *  WithGlobals(v, w).flatMap(f).flatMap(g) =?= WithGlobals(v, w).flatMap(x => f(x).flatMap(g))
+ *  ------------
+ *  WithGlobals(f(v).value, w ∪ f(v).globalVarNames).flatMap(g) =?= ...
+ *  -----------------------------------------------------------
+ *  WithGlobals(g(f(v).value).value, (w ∪ f(v).globalVarNames) ∪ g(f(v).value).globalVarNames) =?= ...
+ *
+ *  ... =?= WithGlobals(v, w).flatMap(x => f(x).flatMap(g))
+ *          -----------------------------------------------
+ *  ... =?= WithGlobals((x => f(x).flatMap(g))(v).value, w ∪ (x => f(x).flatMap(g))(v).globalVarNames)
+ *                      -------------------------------      ----------------------------------------
+ *
+ *  Aside:
+ *    (x => f(x).flatMap(g))(v)
+ *    === f(v).flatMap(g)
+ *    === WithGlobals(g(f(v).value).value, f(v).globalVarNames ∪ g(f(v).value).globalVarNames)
+ *
+ *  ... =?= WithGlobals(g(f(v).value).value, w ∪ (f(v).globalVarNames ∪ g(f(v).value).globalVarNames))
```

Comment:
Hahahaha, ok. Consider adding a clarifying comment that it is not really necessary to understand all the m-word stuff to understand the code (I'm absolutely fine with leaving the proof, personally I hate when people just call "everything that has `map` and `flatMap`" monad :P).